lib/argon2/password.rb

```diff
@@ -0,0 +1,224 @@
+      #    Argon2::Password.create(password, secret: pepper)
+      #    Argon2::Password.create(password, m_cost: 17, secret: pepper)
+      #
+      # Currently available options:
+      #
+      # * :t_cost
+      # * :m_cost
+      # * :secret
+      #
+      def create(password, options = {})
+        raise Argon2::Errors::InvalidPassword unless password.is_a?(String)
+
+        t_cost = options[:t_cost] || DEFAULT_T_COST
+        m_cost = options[:m_cost] || DEFAULT_M_COST
+
+        if t_cost < MIN_T_COST || t_cost > MAX_T_COST
+          raise Argon2::Errors::InvalidTCost
+        end
+
+        if m_cost < MIN_M_COST || m_cost > MAX_M_COST
+          raise Argon2::Errors::InvalidMCost
+        end
+
+        # TODO: Add support for changing the p_cost
+
+        salt = options[:salt_do_not_supply] || Engine.saltgen
+        secret = options[:secret]
+
+        Argon2::Password.new(
+          Argon2::Engine.hash_argon2id_encode(
+            password, salt, t_cost, m_cost, secret
+          )
+        )
+      end
+
+      ##
+      # Regex to validate if the provided String is a valid Argon2 hash output.
+      #
+      # Supports 1 and argon2id formats.
+      #
+      def valid_hash?(digest)
+        /^\$argon2(id?|d).{,113}/ =~ digest
+      end
+
+      ##
+      # Takes a password, Argon2 hash, and optionally a secret, then uses the
+      # Argon2 C Library to verify if they match.
+      #
+      # Also accepts passing another Argon2::Password instance as the password,
+      # in which case it will compare the final Argon2 hash for each against
+      # each other.
+      #
+      # Usage:
+      #
+      #    Argon2::Password.verify_password(password, argon2_hash)
+      #    Argon2::Password.verify_password(password, argon2_hash, secret)
+      #
+      def verify_password(password, digest, secret = nil)
+        digest = digest.to_s
+        if password.is_a?(Argon2::Password)
+          password == Argon2::Password.new(digest)
+        else
+          Argon2::Engine.argon2_verify(password, digest, secret)
+        end
+      end
+    end
+
+    ######################
+    ## Instance Methods ##
+    ######################
+
+    ##
+    # Initialize an Argon2::Password instance using any valid Argon2 digest.
+    #
+    def initialize(digest)
+      digest = digest.to_s
+      if valid_hash?(digest)
+        # Split the digest into its component pieces
+        split_digest = split_hash(digest)
+        # Assign each piece to the Argon2::Password instance
+        @digest   = digest
+        @variant  = split_digest[:variant]
+        @version  = split_digest[:version]
+        @t_cost   = split_digest[:t_cost]
+        @m_cost   = split_digest[:m_cost]
+        @p_cost   = split_digest[:p_cost]
+        @salt     = split_digest[:salt]
+        @checksum = split_digest[:checksum]
+        # The return type is ignored by Object.new, this is provided purely for
+        # return type safety (rbs).
+        self
```

Comment:
Not totally sure if this is necessary, with the way ruby exposes `new` / `initialize`, you typically never touch the return value of `initialize` because `new` does that for you.

Added it to be safe, especially considering it doesn't make much sense for the checksum to be the return type regardless of type safety, and even if you don't really touch it in normal circumstances.

Reply:
Looks like initialize is intended to be a void function. Updated the method signature and removed the self.